### PR TITLE
[Bugfix] Fix HunyuanOCR cross-image contamination in batch processing

### PR DESCRIPTION
**Pull request by anker-c2**
This fixes a bug where differently sized images in a batch would have their features / chunks mixed.

The fix processes each image independently one by one through the transformer layers by splitting, processing, and concatenating per layer.

Fixes #30342

## Purpose
- Fix https://github.com/vllm-project/vllm/issues/30342
- Changes the VisionTransformer such that each image in a bad gets processed independent per Layer
- Done like this instead of implementing a mask like discussed in https://github.com/Tencent-Hunyuan/HunyuanOCR/issues/60 for simplicity. This way no change to attention needed. 
- Split / merge kept inside loop for consistency, could be moved outside. 

## Test Plan
- Ran the changes on a batch of ~900 independent images. 
- Same isolated environment reproduced the same issue in the same resulting chunks three times. 
- Additional test on a minimal example would be appreciated, since i was not yet able to reproduce the behavior on public files. 

## Test Result
- I was not able to reproduce the issue after the fix. All inspected chunks were extracted as expected / found no loss in precision
- Speed seems to be slightly improved as a bonus in some cases (depending on the batch configuration) since less attention calculations should be required at times
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/hunyuan_vision.py [hunk 1]
```diff
@@ -515,8 +516,14 @@ def forward(
...
-        for layer_num, layer in enumerate(self.layers):
-            hidden_states = layer(hidden_states)
+
+        # build per-image lengths once
+        split_lengths = [int(h) * int(w) for (_, h, w) in grid_thw]
+        for layer in self.layers:
+            # hidden_states: (1, T_total, D)
+            parts = hidden_states.split(split_lengths, dim=1)  # list of (1, L_i, D)
+            parts = [layer(p) for p in parts]
+            hidden_states = torch.cat(parts, dim=1)
```
Comment (Isotr0py) [inline vllm/model_executor/models/hunyuan_vision.py:526]: Hmmm, IIUC, the root issue is that we need a docment attention mask for HunyuanOCR's ViT, right?
image-29

Comment (anker-c2) [reply vllm/model_executor/models/hunyuan_vision.py:526]: Right, the root issue seems to be that the image patches get concat together without regard for the size (and therefore no way to separate after) and then get falsely altered across img boundaries during attention. 
A document mask would fix the issue, but i am not sure if it's needed since it would require a lot more changes. 

Imo just processing the images independently per layer like it's done here, instead of building one big attention matrix should achieve the same effect. 

I am not sure about the efficiency of a mask, since i am not well versed in GPU magic: If the attention matrix gets masked, do the extra (useless) computations of the masked fields get optimized away? If so it should be equivalent, but if not shouldn't there be a performance gain computing over multiple (potentionally small) attention matrices depending on the composition of the batch (in regards to the sizes of the included images)?

Either way it should work.

Comment (Isotr0py) [reply vllm/model_executor/models/hunyuan_vision.py:526]: > If the attention matrix gets masked, do the extra (useless) computations of the masked fields get optimized away?

Yes, at least for FlashAttention, it only calculates blocks with valid attention mask (red and yellow area), and skip calculation for invalid mask (blue area). 

95c1afb773cd933c999b53a73ccc8a7f


> but if not shouldn't there be a performance gain computing over multiple (potentionally small) attention matrices depending on the composition of the batch (in regards to the sizes of the included images)?

Not really, launching CUDA kernel can also have overheads, so if there are lots of images, processing them one by one can cause significant overheads for kernel launching as well.

BTW, we're refactoring `MultiHeadAttention`/`MMEncoderAttention` to support passing attention mask. (https://github.com/vllm-project/vllm/pull/30125/files/626d3bba4f8fec51dafe34c0a1089f4691f44d98..628dbfabfac4e978ccf374a544e841cbfef77359?notification_referrer_id=NT_kwDOAncmpLQyMDk3Mzk3NzMzMDo0MTM2MzEwOA), I think we can fix this issue after that PR through passing correct attention mask. :)

Comment (anker-c2) [reply vllm/model_executor/models/hunyuan_vision.py:526]: I see, in that case disregard. Thanks a lot for taking the time to explain it! 

Is there a rough guess on the timeline regarding Mask support feature? I can't estimate the current progress. 
Asking in case it might make sense to employ this as a workaround so that the model at least isn't broken before it's being fixed thoroughly with masking. The current behavior is failing silently, which is a bit of a pain. 
Speedwise it's still slightly faster than the current implementation.

Cheers.

Comment (vince62s) [reply vllm/model_executor/models/hunyuan_vision.py:526]: Hello,
I think this is not the only issue.
The model uses xdrope on the first step=0 pass in the language model.
But this part of the code is broken, it does not support batches:
https://github.com/vllm-project/vllm/blame/main/vllm/transformers_utils/processors/hunyuan_vl.py#L92-L132

**Review by Isotr0py — APPROVED** [review]
> Hello,
I think this is not the only issue.
The model uses xdrope on the first step=0 pass in the language model.
But this part of the code is broken, it does not support batches:
https://github.com/vllm-project/vllm/blame/main/vllm/transformers_utils/processors/hunyuan_vl.py#L92-L132

I see, let's merge this one as workaround for now then.
